qualtran/surface_code/azure_cost_model_test.py

```diff
@@ -0,0 +1,176 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import numpy as np
+import pytest
+from attrs import frozen
+
+from qualtran.surface_code import azure_cost_model
+from qualtran.surface_code.algorithm_specs import AlgorithmSpecs
+from qualtran.surface_code.physical_parameters import BeverlandEtAl as physical_parameters
+from qualtran.surface_code.quantum_error_correction_scheme import GateBasedSurfaceCode
+from qualtran.surface_code.rotation_cost_model import BeverlandEtAl as rotation_cost_model
+from qualtran.surface_code.t_factory import TFactory
+
+
+@frozen
+class Test:
+    alg: AlgorithmSpecs
+    q_alg: float
+
+    error_budget: float
+    c_min: float
+
+    time_steps: float
+    code_distance: float
+
+    t_states: float
+
+
+_TESTS = [
+    Test(
+        alg=AlgorithmSpecs(
+            algorithm_qubits=100,
+            rotation_gates=30100,
+            measurements=14 * 10**5,
+            rotation_circuit_depth=501,
+        ),
+        q_alg=230,
+        error_budget=1e-3,
+        c_min=1.5e6,
+        time_steps=1.5e5,
+        code_distance=9,
+        t_states=6e5,
+    ),
+    Test(
+        alg=AlgorithmSpecs(
+            algorithm_qubits=1318,
+            rotation_gates=2.06e8,
+            measurements=1.37e9,
+            rotation_circuit_depth=2.05e8,
+            toffoli_gates=1.35e11,
+            t_gates=5.53e7,
+        ),
+        q_alg=2740,
+        error_budget=1e-2,
+        c_min=4.1e11,
+        time_steps=4.1e11,
+        code_distance=17,
+        t_states=5.44e11,
+    ),
+    Test(
+        alg=AlgorithmSpecs(
+            algorithm_qubits=12581,
+            rotation_gates=12,
+            measurements=1.08e9,
+            rotation_circuit_depth=12,
+            toffoli_gates=3.73e10,
+            t_gates=12,
+        ),
+        q_alg=25481,
+        error_budget=1 / 3,
+        c_min=1.23e11,
+        time_steps=1.23e10,
+        code_distance=13,
+        t_states=1.49e11,
+    ),
+]
+
+
+@pytest.mark.parametrize('test', _TESTS)
+def test_logical_qubits(test: Test):
```

Comment:
instead of using `pytest.mark.parameterize` this calls for a pytest test fixture